openrouteservice/src/main/java/org/os_concavehull/ConcaveHull.java

```diff
@@ -0,0 +1,1037 @@
+									} else if (on == sn.getPrev()) { // trim sn
+										HullEdgeCir oe = nodeEdgeMap.get(on);
+										edgeIdx.remove(oe); // se popped out already
+										coordNodeMap.remove(nodeS);
+										nodeEdgeMap.remove(sn);
+										nodeEdgeMap.remove(on);
+										hull.remove(sn); //on now connected to en
+										//
+										HullEdgeCir oeNew = new HullEdgeCir(on);
+										nodeEdgeMap.put(on, oeNew);
+										edgeIdx.add(oeNew);
+										if (keepLineSeg) {
+											Coordinate[] lsCoords = new Coordinate[2];
+											lsCoords[0] = nodeO;
+											lsCoords[1] = nodeS;
+											rltLS.add(gf.createLineString(lsCoords));
+										}
+										rltPt.add(gf.createPoint(sn.getObj()));
+									} else {// split
+										addHull = false;
+										// split
+										DLCirList hull2 = hull.split(sn, en, on);
+										hulls.add(hull);
+										hulls.add(hull2);
+										//
+										break;
+									}
+								} else {// multiparts not allowed
+									// don't do anything
+								}
+								//
+							} else { // normal digging
+								//
+								nodeEdgeMap.remove(sn);
+								DLNode<Coordinate> on = new DLNode<Coordinate>(nodeO);
+								coordNodeMap.put(nodeO, on);
+								hull.addAfter(sn, on);
+								//
+								HullEdgeCir seNew = new HullEdgeCir(sn);
+								HullEdgeCir oe = new HullEdgeCir(on);
+								edgeIdx.add(seNew);
+								edgeIdx.add(oe);
+								nodeEdgeMap.put(sn, seNew);
+								nodeEdgeMap.put(on, oe);
+							}
+						} else {// 3 vertices only, clapse to a line segement, may be saved separately if needed?
+							addHull = false;
+							break;
+						}
+					}
+				}// end while(!edgeIdx.isEmtpy())
+				// 
+				if (addHull) {
+					// add hull to result
+					rltHulls.add(hull);
+				}
+				// clear indices
+				coordNodeMap.clear();
+				nodeEdgeMap.clear();
+				edgeIdx.clear();
+			}
+			Vector<Geometry> rtn = new Vector<Geometry>(hulls.size() + rltLS.size() + rltPt.size());
+			for (DLCirList<Coordinate> h : rltHulls) {
+				Coordinate[] coords = new Coordinate[h.size() + 1];
+				int cnt = 0;
+				DLNode<Coordinate> node = h.getNode();
+				do {
+					Coordinate coord = node.getObj();
+					coords[cnt++] = coord;
+					node = node.getNext();
+				} while (node != h.getNode());
+				coords[cnt] = new Coordinate(coords[0]);
+				Geometry geom = gf.createPolygon(coords);
+				rtn.add(geom);
+			}
+			rtn.addAll(rltLS);
+			rtn.addAll(rltPt);
+			//
+			return rtn;
+		} else {
+			return null;
+		}
+	}    //
+
+	/** Experimental Concave hull construction with an extra metric to control the order of boundary edge "digging"
+	 * @param triChecker
+	 * @param m TriangleMetric for controlling the order of digging (default is the length of the boundary edge)
+	 * @param allowMultiParts
+	 * @param keepLineSeg
+	 * @return
+	 */
+	public Collection<Geometry> getConcaveHullMetric(TriangleChecker triChecker, TriangleMetric m, boolean allowMultiParts, boolean keepLineSeg) {

```

Comment:
Please refactor this method into several shorter ones.

Reply:
Removed